dll/cpl/desk/devsett.c

```diff
@@ -515,6 +560,76 @@ pCDevSettings_Initialize(PCDevSettings This,
         RegCloseKey(hKey);
     }
 
+    /* Initialize display modes */
+    ZeroMemory(&devmode, sizeof(devmode));
+    devmode.dmSize = (WORD)sizeof(devmode);
+    while (EnumDisplaySettingsExW(This->pDisplayDevice, i, &devmode, EDS_RAWMODE))
+    {
+        dwSize = devmode.dmSize + devmode.dmDriverExtra;
+        PDEVMODEW pDevMode = LocalAlloc(LMEM_FIXED, dwSize);
+        PDEVMODEW * DevModesNew = NULL;
+
+        if (pDevMode)
+        {
+            CopyMemory(pDevMode,
+                       &devmode,
+                       dwSize);
+
+            dwSize = (This->nDevModes + 1) * sizeof(pDevMode);
+            DevModesNew = LocalAlloc(LMEM_FIXED, dwSize);
+            if (DevModesNew)
+            {
+                if (This->DevModes)
+                {
+                    CopyMemory(DevModesNew,
+                               This->DevModes,
+                               This->nDevModes * sizeof(pDevMode));
+
+                    LocalFree(This->DevModes);
+                }
+
+                This->DevModes = DevModesNew;
+                This->DevModes[This->nDevModes++] = pDevMode;
+            }
+            else
+            {
+                DPRINT1("LocalAlloc failed to allocate %d bytes\n", dwSize);
+                return E_OUTOFMEMORY;
+            }
+        }
+        else
+        {
+            DPRINT1("LocalAlloc failed to allocate %d bytes\n", dwSize);
+            return E_OUTOFMEMORY;
+        }
+
+        devmode.dmDriverExtra = 0;
+        i++;
+    }
+
+    /* FIXME: Remove duplicated modes */
+
+    /* Get current display mode */
+    ZeroMemory(&devmode, sizeof(devmode));
+    devmode.dmSize = (WORD)sizeof(devmode);
+    if (EnumDisplaySettingsExW(This->pDisplayDevice, ENUM_CURRENT_SETTINGS, &devmode, 0))
+    {
+        for (i = 0; i < This->nDevModes; i++)
+        {
+            if (!This->DevModes[i])
+                continue;
+
+            if (((This->DevModes[i]->dmFields & DM_PELSWIDTH) && devmode.dmPelsWidth == This->DevModes[i]->dmPelsWidth) &&
```

Comment:
That's a whole lot of dereferencing `This->DevModes[i]` again and again, it might make sense to store this pointer once and not look it up every time?

Same for a few other cases down below.

Reply:
Yep, that's good idea. Also this whole file looks like a reconstruction of a C++ class, but we're using C here again :smile:

Reply:
yeah, some people like to do the work of a compiler manually, 300000 times.